Feature: Arterytek option bytes and more part IDs

ALTracer opened the pull request:
## Detailed description

* This is a new feature.
* Currently there's no easy way to extend SRAM of AT32F403A via `monitor option` commands because EOPB0 is out of reach for the normal 16-byte driver.
* The PR solves it by conditionally bumping the option byte ranges.

Past first commit see my work of adding yet more part ID lists for other families in AT32F4 lineup. This is probably getting unmanageable, so I started incorporating F_SIZE and "Product ID" readouts from UID (in their DESig analog). If this is fine, e.g. does not result in problems for RDP1-enabled targets, then all such lists can be dropped or gated under ENABLE_DEBUG to not hog precious BMF flash -- it only needs to know flash parameters (banks, sector size, total size), not exact package & pinout. I may move this to a separate PR (and a split translation unit) as no users have probably asked for F421, F413 or F425. F423 and F402/F405 are plain unimplemented (F405 with USB HS PHY would be nice to have though). I have none of these chips to test on.

## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (see [Building the firmware](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-firmware)) ...and overflows flash
* [x] It builds as BMDA (see [Building the BMDA](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-app))
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues

Allows operating Arterytek User System Data, including Flash Access Protection enable/disable ops.

dragonmux commented on src/target/stm32f1.c:
```diff
@@ -820,16 +955,22 @@ static bool stm32f1_option_write_erased(
 	return status == SR_PROG_ERROR;
 }
 
+#define STM32F1_OB_COUNT 8U
+#define AT32F4_OB_COUNT 24U
```
All `#define`'s should go at the top of the file after the `#include` block please.

ALTracer replied:
Moved up.

dragonmux reviewed (commented):
Please can you rebase this on `main` and run `clang-format`. We'll do a proper review at that point rather than just a quick cursory as we're anticipating it's pretty close to already being mergable.

Apologies this one slipped through the cracks and we failed to review it at all till now.

ALTracer commented:
Rebased and clang-formatted.
I only wanted 48 option bytes for AT32F403A (especially RDP1 and EOPB0). Even that impl is bad -- it would affect GD32F1 etc. because of TOPT_32BIT criteria. The stm32f1-generic flash driver does not record option byte region size. If you have any better ideas -- I'll try to edit them in for a minor but effective change (no large driver-scope refactoring).
I would rather migrate/move the rest of chips/targets into at32f43x.c sometime in the future, but that should be made more generic as well to handle that. And it'd shrink the stm32f1.o (always compiled).

dragonmux commented on src/target/stm32f1.c:
```diff
@@ -418,11 +428,15 @@ static bool at32f40_detect(target_s *target, const uint16_t part_id)
 	target->target_options |= STM32F1_TOPT_32BIT_WRITES;
 	target->mass_erase = stm32f1_mass_erase;
 
+	// AT32F403A/F407 have 48 bytes of User System Data
```
This is a useful tidbit to know but we don't see how it relates to the line after it or how it impacts things. Please flesh this out as a comment perhaps around line 424 as part facts and how they reflect onto the implementation.

dragonmux replied:
Ah, okie - in that case perhaps prefix this with `NB: ` or so to indicate that you're delivering some extra information?

ALTracer replied:
Expanded oneline comment to carry more meaning.

dragonmux commented on src/target/stm32f1.c:
```diff
@@ -457,11 +471,112 @@ static bool at32f41_detect(target_s *target, const uint16_t part_id)
+
+static bool at32f425_detect(target_s *target, const uint16_t part_id)
+{
+	const uint16_t flash_size = target_mem32_read16(target, AT32F4x_FLASHSIZE);
```
This doesn't appear to check any ID codes before going "yeah, sure, it's a AT32F425"?

ALTracer replied:
Correct, because the checking was done by its caller against `0x5009_2???` series. This function is implemented simpler because my theory is F_SIZE might be enough for BMD purposes.

dragonmux commented on src/target/stm32f1.c:
```diff
@@ -473,11 +588,33 @@ bool at32f40x_probe(target_s *target)
+	if (series == AT32F425_SERIES && project_id == 0x0fU)
+		return at32f425_detect(target, part_id);
+	if ((series == AT32F423_SERIES_256KB || series == AT32F423_SERIES_128KB || series == AT32F423_SERIES_64KB) &&
+		(project_id == 0x12U))
```
The extra parens here are superfluous and can be dropped - same on line 607 for the clause after the `&&`. They suppress the GCC warning about accidental assignment that we want here.

ALTracer replied:
Removed two pairs of parens for project_id.

dragonmux reviewed (commented):
There are a couple of remaining review notes to deal with, and once they've been sorted we're happy for this to merge.

ALTracer commented:
While I don't think it ends good for GD32 in current form, I'd like to ask to `if(!strcmp(target->driver, "AT32"))` instead of TOPT so both would work as expected.
Rebased away the addition of AT32F405/F402 (HS USB) because their FPEC is at 0x40023C00 like F435/F437, not 0x40022000 as this driver is hardcoded to work on. Same for AT32F423 (three series, FPEC at incompatible offset, unsupported here).

dragonmux commented:
Once the single outstanding review note has been sorted, we're happy to get this merged. This is looking really good.

ALTracer commented:
Replaced the option byte size dispatch check from TOPT 32-bit writes (shared with Gigadevices) to direct strncmp. This triggers correctly, 48 bytes (24 halfwords) for AT32F403A/407, but still 16 bytes (8 halfwords) for GD32F103 etc.
Did not register cmd_option for older F403 or newer F413, as I cannot test on them. Did not register it for F415 either, because it has 1024 not 48 bytes.
Added a few more families very likely compatible with STM32F1 flash driver. They all have UID/F_SIZE unreadable under RDP1, so need part ID tables. I started to notice a pattern "part ID <=> flash density", but the optimization would be handled in a follow-up PR, if ever.
Please review again.

dragonmux commented on src/target/stm32f1.c:
```diff
@@ -459,11 +519,143 @@ static bool at32f41_detect(target_s *target, const uint16_t part_id)
+
+static bool at32f425_detect(target_s *target, const uint16_t part_id)
+{
+#if 1
```
This preprocessor logic makes the `#else` part moot - please don't include dead code if at all possible. If what's needed can be achieved with just this 16-bit register read, then let's do that instead of having a large part ID switch-case.

ALTracer replied:
No, there is a major problem of us not knowing flash size when RDP1 is active. Part IDs allow that. I would rather delete the easy two-liner and postpone it to the refactor, relying on part ID tables like all other routines (this table is thankfully small). We previously argued that read-locked flash should not be added anyways, just the option byte erase handler; but it's not implemented for 512 byte region.

ALTracer replied:
Inverted macro to gate the easy path and added TODO comments.

dragonmux reviewed (changes requested):
Having properly re-reviewed this, there's just one thing we found. Please address it and rebase on `main` then we'll approve this for merge.

dragonmux reviewed (approved):
LGTM, all items addressed. Merging. Thank you for the contribution!
